Custom Elements ADR (#981)
primer/components: contributor-docs/adrs/adr-001-custom-elements.md

- November 27th 2019: [Usage of custom elements in IIE project](https://github.com/github/issues-index-experiment/pull/133)
- November 27th 2019: [Styling custom elements with styled components](https://github.slack.com/archives/CMZ4DC9BL/p1574883198055600)
- June 9th 2020: [Discussion in Slack about using skatejs for SSR](https://github.slack.com/archives/C0ER2LCG2/p1591707104461300)
- November 10th 2020: [Replace dialog with details-dialog](https://github.com/primer/components/issues/907)

## Context
Throughout the last few years folks from the Design Infrastructure, Web Systems & UI Platform teams have discussed the idea of using custom elements for behaviors in Primer React Components. The main goal of using custom elements in PRC is to be able to author behaviors once and reuse them in any framework. Several experiments have been conducted which are listed above.

## Assumptions
De-duplication is not our highest or only priority. Attempts at de-duplication must be weighed against changes to the maintainer, developer, and customer experience.

## Findings

### Developer Experience Regressions

> - Rendering custom elements in SSR applications is a bit tricky and requires a third party framework such as [@skatejs/ssr](https://github.com/skatejs/skatejs/tree/master/packages/ssr). This is a hit to developer experience because it would require consumers of PRC to set up the service, PRC won’t work right out of the box (in applications using SSR, such as all of our Doctocat sites) without it if we use custom elements in the library.

Reviewer: Nit: A third-party framework isn’t necessarily required; I wrote a ~50 line [function](https://github.com/github/ui-eng-framework-survey/blob/6a8ec7c2080146c7495c025500c96cc7287cbb20/utils.ts#L50-L111) for rendering custom elements server-side for `github/ui-eng-framework-survey`

Reviewer 2: Custom Elements are a fundamentally different design pattern to React, which I think is what is causing confusion here. Additionally the term "SSR" means many different things, but they're conflated into one with React.

To borrow a term used in the React ecosystem, React components are designed with _isomorphism_ as an ideal while Custom Elements are the opposite. The `class extends HTMLElement {}` part of a custom element is designed to _only run in the browser execution environment_. This means They shouldn't be imported and run on the server. You can try to do that by loading a DOM library like `jsdom` into the server, but it's attempting to solve a problem which shouldn't be solved.

This doesn't mean Custom Elements cannot be rendered server side, of course. The HTML part of Custom Elements is just that, HTML. It'll work anywhere that can output a string. Bash can be used as a server side language and output Custom Elements, just like React, just like Go, just like C, just like Rails. HTML is the primitive which is _platform agnostic_ and the server emits HTML tags that match the names of the custom elements, and the Browser will combine the JS logic to the HTML tag. This is progressive enhancement. It means you can ship custom elements that don't require JS to function.

A big part of the "SSR story" is rendering subtrees. Again, because Custom Elements are platform agnostic they leave rendering of subtrees to the user. For example our `<details-dialog>` component expects to be rendered as a sub-element of `<details>`, so the _implicit_ expectation is that in whatever server side language you're using, you write a component that does that wiring. In the case of the monolith we use ViewComponents for that. React could be used too. This grants flexibility in the Server Side implementation without tying you to JS. Does it mean writing two implementations for a component? No (unless you're trying hard to do that). Custom Elements strongly encourage progressive enhancement, and technology agnostic to servers. The JS _shouldn't_ be doing hydration, subtreets, etc, if you can avoid it.

If you _want_ to render SubTrees in Custom Elements, it is possible using Shadow DOM. Server side rendering of ShadowDOM is still nascent, but the [Declarative Shadow DOM spec](https://web.dev/declarative-shadow-dom/) will solve this. It's not shipping in browsers yet, but it can be polyfilled with a few lines of JS. [Catalyst already ships a polyfill for declarative shadowDOM templates that is 11LOC](https://github.com/github/catalyst/blob/main/src/auto-shadow-root.ts). Declarative Shadow DOM is again HTML only, so is not coupled to JS or any framework, React can output Declarative ShadowDOM contents similar to how React might write a component itself:

```js
const ReactProgressBar = ({progress, bg, theme, ...rest}) => {
  return (
    <ProgressContainer theme={theme} {...rest}>
      <Bar progress={progress} bg={bg} theme={theme} />
    </ProgressContainer>
  )
}

const ShadowDOMProgressPar = ({progress, bg, theme, ...rest}) => {
 return (
    <progress-container theme={theme} {...rest}>
      <template shadowroot="open">
        <bar-element progress={progress} bg={bg} theme={theme} />
      </template>
    </progress-container>
  )
}

> - You cannot style custom elements with styled-components[^1]. This means that if a component wants to use a custom element to get behaviors and you also want to style that component, you must use another wrapper div to apply styles. This feels awkward and introduces unnecessary markup to the DOM.

Reviewer 2: Looks like a bug in Styled Components, https://github.com/styled-components/styled-components/pull/2819, will be fixed in the next release.

```jsx
export const ClipboardCopy = styled('clipboard-copy')
```

`;

const App = () => {
  return (
   <>
    <StyledClipboardCopyWrapper>
     <ClipboardCopy />
    </StyledClipboardCopyWrapper>
    <span id="span">Content from span tag.</span>
  </>
 );
};
```

View in CodeSandbox: https://codesandbox.io/s/demo-styling-custom-element-2-xt9g9?file=/src/index.tsx

This also means we would need to be careful to never export a component with a custom element as the top level node.

### Incompatibility with some React tools
Some of our GitHub custom elements such as details-dialog and details-menu make assumptions about the DOM tree. For example, details-dialog expects a `details` element to wrap the custom element and uses this assumption[^2] to determine whether or not clicks are happening inside or outside of the dialog and closes the dialog if the click happened outside of the dialog. This makes sense in most cases and is a nice way of enforcing proper usage of the details element, but breaks down when used with [React Portals](https://reactjs.org/docs/portals.html) which are often used to ensure menus are displayed correctly in cases where a parent has an overflow: hidden applied to it, or incompatible z-index.

### Extensibility

> Building behaviors in React Hooks gives us the ability to provide things like state and state change hooks to the consumer of the component. This allows the user to build on additional behaviors to the component based on the state or other variables provided to the component consumer. Doing the same with custom elements would require listening to events on the document[^3] and reacting to them. This is certainly do-able, but goes against some of the foundational principles of React (reacting to changes in the DOM vs changes in React state).

Reviewer 2: The equivalent of state in React is attributes in Custom Elements. Custom Elements can be configured by attributes and changes to attributes can be reacted to without event binding. Event binding is a way for components to communicate up the tree, and is conceptually perhaps similar to hooks in general.

Author: Yes, that's basically what I'm saying here in this paragraph. The point is that if we use custom elements in our react library we will have to rely on attributes to pass around state + listening to changes to those attributes versus keeping state change management within the realm of the React ecosystem.

### Organizational Overhead

> - GitHub’s custom elements are all managed in different repos which introduces more maintenance overhead.

Reviewer 2: What additional maintenance overhead does this create?

Author: 1. You'd need to `npm link` while developing if you want to test changes out with the presentational components themselves instead of making changes and seeing updates instantly. `npm link` usually doesn't work well with hot module reloading either.
2. You'd need to draft & publish releases to both libraries every time you want to update the behavior
3. If the behaviors are shared between .com and PRC you'd need to do careful testing in both environments to make sure that changes don't create any regressions. That greatly widens the context that engineers need to keep in mind every time a change is made.

Reviewer 2: Let's enumerate these points in the ADR:

```suggestion
- GitHub’s custom elements are all managed in different repos which introduces more maintenance overhead.

  - You'd need to npm link while developing if you want to test changes out with the presentational components themselves instead of making changes and seeing updates instantly. npm link usually doesn't work well with hot module reloading either.
  - You'd need to draft & publish releases to both libraries every time you want to update the behavior
  - If the behaviors are shared between .com and PRC you'd need to do careful testing in both environments to make sure that changes don't create any regressions. That greatly widens the context that engineers need to keep in mind every time a change is made.
```

> - Reacting to changes will take a bit more time, as we’ll need to wait for changes to be made to the custom elements until we can update Primer React Components (versus having behavior Hooks in PRC that we can update at our leisure).

Reviewer 2: Could you expand on this? I don't see the issue here. Why can't you update custom elements at your leisure?

Author: See my comment above

Reviewer 2: ```suggestion
- Reacting to changes will take a bit more time, as we’ll need to orchestrate releases between custom elements and Primer React Components - as opposed to having beheaviors already present in PRC which can be versioned in lockstep.
```

### Other
- The custom element & web component API progress slower than React due to changes needing to go through the whatwq standards process.

### Risks of not switching to custom elements for behaviors
- We spend extra time building behaviors in React that have already been built in our [custom elements library](https://github.github.io/web-systems-documentation/#custom-elements).
  - There are currently 19 behaviors/components listed on the custom elements documentation site. Several of these we have already implemented in React in either PRC, Doctocat, or other React applications at GitHub which can be upstreamed (details-dialog, details-menu, clipboard-copy, text-expander, autocomplete, task-list via drag and drop hooks, tab-container, text-expander).
- We decide not to invest further in React at GitHub and have wasted time we could have spent building more custom elements.
  - This seems unlikely as there seems to be clear consensus that we will continue to build more and more highly interactive products.
- The React library is abandoned and becomes obsolete

>   - This is a risk with any technology that we may use, seems highly unlikely in the near term, and is also a risk for custom elements.

Reviewer 2: This isn't an existential risk to Custom Elements. Browsers can't deprecate API; the APIs will remain indefinitely.

Author: Yes, but it's possible that further work on them could be dropped

Reviewer: From @keithamus in https://github.com/primer/components/pull/981#discussion_r564714720:
> …Browsers can't deprecate API…

Could you please clarify this, @keithamus—are you talking about [Custom Elements APIs](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) specifically?

I’m not sure this holds as a general principle: For example, AppCache’s [MDN page](https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache) begins with the following warnings:

> **Do not use this _application cache_ feature!** It is [in the process of being removed from the Web platform](https://html.spec.whatwg.org/multipage/browsers.html#offline).
> 
> - From Firefox 84 it has been removed ([bug 1619673](https://bugzilla.mozilla.org/show_bug.cgi?id=1619673)). It is also planned for removal in Chomium 90, and is deprecated in Safari.

> 👎 **Deprecated**
> This feature is no longer recommended. Though some browsers might still support it, it may have already been removed from the relevant web standards, may be in the process of being dropped, or may only be kept for compatibility purposes. Avoid using it, and update existing code if possible; see the [compatibility table](https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache#Browser_compatibility) at the bottom of this page to guide your decision. Be aware that this feature may cease to work at any time.

Reviewer 2: `AppCache` is an extremely rare case where browsers were able to deprecate an API. However it serves a good example of just how carefully APIs get deprecated, if ever. AppCache was a good candidate for deprecation as it was overly complicated, had serious architectural concerns, hampered development of new specs, not quite fit for purpose, and only 1.5% of the web used it. Most use cases used the declarative HTML attribute which meant disabling the API would have a minimal effect. Even so the API was originally marked for deprecation back in 2015, and it took several years for browsers to begin marking it as deprecated, with Chrome removing it from non-HTTPS sites in 2018, and shipping an origin trial removing it from HTTPS sites in 2020. As of writing Chrome still hasn’t deprecated AppCache - it’s due for removal in Chrome 90. In addition to this the Chrome team has provided a drop in library for those who are unable to remove the feature from their sites. So AppCache, a feature used by a small % of the web, has taken 6 years to get close to being deprecated, with browsers offering support for migrating throughout. 

Another example of a deprecated api would be Custom Elements v0 of which GitHub was a customer. This had a similar story: small amounts of usage, preventing new developments; with deprecation announced in 2014 via many deprecation notices, as well as support libraries offered. A paved path was provided to upgrade to the v1 spec and it took around 6 years to finally remove v0, with Chrome 80s release in 2020. GitHub migrated our existing components from v0 to v1, and Google engineers communicated with us to support the transition (see https://github.com/github/time-elements/pull/67 for an example). 

The difference between browsers deprecating APIs and, say, react deprecating APIs, is that the browsers have to maintain compatibility with websites. That's why we still have a `<font>` tag. React can release a major version and we either have to migrate or stick with an older version which may no longe receive support. The web can deprecate APIs but it’s going to be a multi year process and ultimately they’re beholden to web authors will to change.

> - The React library is abandoned and becomes obsolete
>   - This is a risk with any technology that we may use, seems highly unlikely in the near term, and is also a risk for custom elements.

Reviewer 2: ```suggestion
- The React library is abandoned and becomes obsolete
  - This is a risk with any technology that we may use, seems highly unlikely in the near term.
  - While also a possibility for Custom Elements, the track record demonstrates deprecations of Web APIs is extremely rare and has a long deprecation path.
```

- Behaviors in .com using custom elements and behaviors in PRC diverge, leading to an inconsistent experience
  - This is probably the biggest risk we face, but moving to custom elements isn’t necessarily the only or best solution. We should explore other ways of detecting divergence such as integration tests.

> 

Reviewer 2: In addition to these risks there's a large risk of duplicate efforts & engineering hours wasted developing interactions for two different stacks.

Author: This is addressed in the first bullet point

### Decision

> Due to the challenges listed above and our priorities listed in the [Assumptions](#assumptions) section, we are not investing time in building out behaviors with custom elements in our Primer React Components library. Instead, we should spend time expanding coverage using React Hooks and focus on finding other approaches for making sure implementation of behaviors in our different stacks are consistent (such as integration tests).

Reviewer 2: This seems to enumerate the reasons for not implementing behaviours in Web Components but fails to address the reasons why we _should_ implement behaviours using React Hooks.

Implementing behaviours in React Hooks couples them to React. Another option for us is to write functions which return predefined interfaces that can be integrated with _either_ hooks _or_ Web Components. Both Web Component's lifecycle methods and React Hooks are effectively interchangeable injection points (e.g. `useEffect` is analogous to `connectedCallback`). Why would we not explore this avenue further?

Author: > fails to address the reasons why we should implement behaviours using React Hooks

@T-Hugs brought up the point yesterday that we could also test out writing behaviors in vanilla JS to be reused across any UI stack at GitHub. This approach would still face the same organizational & maintenance challenges I've detailed in this ADR.

Reviewer 3: > Another option for us is to write functions which return predefined interfaces that can be integrated with either hooks or Web Components. Both Web Component's lifecycle methods and React Hooks are effectively interchangeable injection points (e.g. useEffect is analogous to connectedCallback). Why would we not explore this avenue further?

I think this is outside the scope of this ADR. I am working on exploring this.

Author: @T-Hugs @keithamus  I might be still missing some context around what you two discussed last week regarding [strategy alignment](https://github.slack.com/archives/C01C60YF4BE/p1611249458000700?thread_ts=1611227360.000200&cid=C01C60YF4BE). From what I gathered it sounded like we were all in alignment with using React Hooks & React state to manage behaviors. Can one of you write up a summary of what you discussed? Behavior work is scheduled to start next week, and we have deadlines relating to big ships later this year so changing strategy now is going to change the epic we've planned quite a bit. Either way, we need to work quickly to get into alignment.
